src/Sylius/Bundle/CartBundle/Controller/CartItemController.php

```diff
@@ -53,49 +39,61 @@ public function addAction(Request $request)
 
-        $eventDispatcher = $this->getEventDispatcher();
+        $itemQuantityModifier = $this->get('sylius.order_item_quantity_modifier');
+        $itemQuantityModifier->modify($newResource, 1);
 
-        try {
-            $item = $this->getResolver()->resolve($emptyItem, $request);
-        } catch (ItemResolvingException $exception) {
-            // Write flash message
-            $eventDispatcher->dispatch(SyliusCartEvents::ITEM_ADD_ERROR, new FlashEvent($exception->getMessage()));
+        $form = $this->resourceFormFactory->create($configuration, $newResource);
 
-            return $this->redirectAfterAdd($configuration);
-        }
+        if ($request->isMethod('POST') && $form->submit($request)->isValid()) {
+            $newResource = $form->getData();
 
-        $event = new CartItemEvent($cart, $item);
+            $event = $this->eventDispatcher->dispatchPreEvent(ResourceActions::CREATE, $configuration, $newResource);
 
-        // Update models
-        $eventDispatcher->dispatch(SyliusCartEvents::ITEM_ADD_INITIALIZE, $event);
-        $eventDispatcher->dispatch(SyliusCartEvents::CART_CHANGE, new GenericEvent($cart));
-        $eventDispatcher->dispatch(SyliusCartEvents::CART_SAVE_INITIALIZE, $event);
+            if ($event->isStopped() && !$configuration->isHtmlRequest()) {
+                throw new HttpException($event->getErrorCode(), $event->getMessage());
+            }
+            if ($event->isStopped()) {
+                $this->flashHelper->addFlashFromEvent($configuration, $event);
 
-        // Write flash message
-        $eventDispatcher->dispatch(SyliusCartEvents::ITEM_ADD_COMPLETED, new FlashEvent());
+                return $this->redirectHandler->redirectToIndex($configuration, $newResource);
+            }
 
-        return $this->redirectAfterAdd($configuration);
-    }
+            $cart = $this->getCurrentCart();
+            $this->resolveCartItem($cart, $newResource);
 
-    /**
-     * Redirect to specific URL or to cart.
-     *
-     * @param Request $request
-     *
-     * @return RedirectResponse
-     */
-    private function redirectAfterAdd(RequestConfiguration $configuration)
-    {
-        $request = $configuration->getRequest();
+            $this->eventDispatcher->dispatchPostEvent(ResourceActions::CREATE, $configuration, $newResource);
+
+            if (!$configuration->isHtmlRequest()) {
+                return $this->viewHandler->handle($configuration, View::create($newResource, Response::HTTP_CREATED));
+            }
+
+            $translatedMessage = $this->translateMessage('sylius.cart.item_add_completed', $this->metadata);
+            $this->addFlash('success', $translatedMessage);
 
-        if ($request->query->has('_redirect_to')) {
-            return $this->redirectHandler->redirect($configuration, $request->query->get('_redirect_to'));
+            $this->recalculateCart($cart);
```

Comment:
This should be done via listener. The addAction should dispatch CART_CHANGE event and listener should handle recalculation.